feat: introduce entites crossId on import process

a-cordier opened the pull request:
https://github.com/gravitee-io/issues/issues/7003

<!-- Storybook placeholder -->
---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-lpbjgguaov.chromatic.com)
<!-- Storybook placeholder end -->
<!-- UI placeholder -->
🚀 CI was able to deploy the build of this PR, so you can now try it directly [here](https://apimnightlywebui24386.blob.core.windows.net/fix-importrefactor/index.html)
_Notes_: The deployed app is linked to the management API of APIM master. (Same login and password as APIM master)
<!-- UI placeholder end -->

Commit:
refactor: remove some useless code from ApiServiceImpl.update

a-cordier committed:
feat(import): add a cross ID to match entities accross environments

If the API definition is imported from another environment we need to match the targeted
entities (API, plans, pages) from the input IDs.

If the API definition holds a cross ID, the entities will be matched using this unique ID.

If the API definition does not hold a cross ID, the matching entity IDs will be matched using
a predictable ID generation based on the target environment ID, the source API ID and the source entity ID

marcambier commented on gravitee-apim-repository/gravitee-apim-repository-jdbc/src/main/java/io/gravitee/repository/jdbc/management/JdbcApiRepository.java:
```diff
@@ -525,4 +526,38 @@ private String convert(ApiCriteria apiCriteria) {
+            LOGGER.debug("JdbcApiRepository.findByEnvironmentIdAndCrossId({}, {}) = {}", environmentId, crossId, result);
+            return result;
+        } catch (final Exception ex) {
+            LOGGER.error("Failed to find api by id:", ex);
```
```suggestion
            LOGGER.error("Failed to find api by environment and crossId", ex);
```

marcambier commented on gravitee-apim-repository/gravitee-apim-repository-jdbc/src/main/java/io/gravitee/repository/jdbc/management/JdbcApiRepository.java:
```diff
@@ -525,4 +526,38 @@ private String convert(ApiCriteria apiCriteria) {
+            return result;
+        } catch (final Exception ex) {
+            LOGGER.error("Failed to find api by id:", ex);
+            throw new TechnicalException("Failed to find api by id", ex);
```
```suggestion
            throw new TechnicalException("Failed to find api by environment and crossId", ex);
```

marcambier commented on gravitee-apim-rest-api/gravitee-apim-rest-api-model/src/main/java/io/gravitee/rest/api/model/api/ApiEntity.java:
```diff
@@ -50,6 +47,9 @@
     @ApiModelProperty(value = "API's uuid.", example = "00f8c9e7-78fc-4907-b8c9-e778fc790750")
     private String id;
 
+    @ApiModelProperty(value = "API's crossId. Identifies API across environments.", example = "00f8c9e7-78fc-4907-b8c9-e778fc790750")
```
```suggestion
    @ApiModelProperty(value = "API's crossId. Identifies API across environments.", example = "df83b2a4-cc3e-3f80-9f0d-c138c106c076")
```

Commit:
feat(management): allow searching API by crossId

Allow searching APIs by crossId via management Rest APIS, using the "crossId" query param.
That will simplify CI/CD usecases.

Commit:
feat(management): add PUT endpoint to import API from definition, without specifying his id

This simplifies using API import endpoints from CI/CD.

As this PUT endpoint doesn't contains API id, we have to check authenticated user permissions manually after recalculating target API id.

Commit:
refactor: remove ApiDuplicatorService.createWithImportedDefinition useless 'user' parameter

phiz71 commented on gravitee-apim-repository/gravitee-apim-repository-mongodb/src/main/java/io/gravitee/repository/mongodb/management/internal/api/ApiMongoRepository.java:
```diff
@@ -25,4 +26,6 @@
 @Repository
-public interface ApiMongoRepository extends MongoRepository<ApiMongo, String>, ApiMongoRepositoryCustom {}
+public interface ApiMongoRepository extends MongoRepository<ApiMongo, String>, ApiMongoRepositoryCustom {
+    Optional<ApiMongo> findByEnvironmentIdAndCrossId(String environmentId, String crossId);
```
With the `@Query`, it would be clearer ;)

marcambier replied:
added it. nice catch

phiz71 commented on gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/converter/PageConverter.java:
```diff
@@ -0,0 +1,72 @@
+        newPage.setSource(pageEntity.getSource());
+        newPage.setType(PageType.valueOf(pageEntity.getType()));
+        newPage.setAttachedMedia(pageEntity.getAttachedMedia());
+        newPage.setVisibility(pageEntity.getVisibility());
```
in the original convert method, Visibility was not set. Is it done on purpose here ?

marcambier replied:
As discussed together together, it sound relevant to map the Page visibility here.

phiz71 commented on gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/converter/PageConverter.java:
```diff
@@ -0,0 +1,72 @@
+        updatePageEntity.setOrder(pageEntity.getOrder());
+        updatePageEntity.setParentId(pageEntity.getParentId());
+        updatePageEntity.setPublished(pageEntity.isPublished());
+        updatePageEntity.setVisibility(pageEntity.getVisibility());
```
in the original convert method, Visibility was not set. Is it done on purpose here ?

marcambier replied:
As discussed together together, it sound relevant to map the Page visibility here.

phiz71 commented on gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/ApiDuplicatorServiceImpl.java:
```diff
@@ -633,79 +610,202 @@ public int hashCode() {
+                () -> recalculateIdsFromDefinitionIds(environmentId, apiJsonNode, urlApiId)
+            );
+
+        return generateEmptyIds(apiJsonNode);
```
```suggestion
        if (!apiJsonNode.hasId() || !apiJsonNode.getId().equals(urlApiId)) {
                findApiByEnvironmentAndCrossId(environmentId, apiJsonNode.getCrossId())
                    .ifPresentOrElse(
                        api -> recalculateIdsFromCrossId(apiJsonNode, api),
                        () -> recalculateIdsFromDefinitionIds(environmentId, apiJsonNode, urlApiId)
                    );
        }
        return generateEmptyIds(apiJsonNode);
```

phiz71 commented on gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/rest/api/service/ApiDuplicatorService_CreateWithDefinitionTest.java:
```diff
@@ -453,8 +444,8 @@ public void shouldCreateImportApiWithPlans() throws IOException {
 
-        String plan1newId = "b1761e88-e346-34f5-ac63-f0ba0ccf8869";
-        String plan2newId = "14ad0cbe-fbaa-3161-acb0-add9b500fc03";
+        String plan1newId = "310f7dd5-71b9-3bdd-b74d-4b76bc71c5b1";
+        String plan2newId = "ed7bef56-b9ec-3c0a-9d21-02e046df429c";
```
why ?

phiz71 commented on gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/rest/api/service/impl/ApiDuplicatorServiceImplTest.java:
```diff
@@ -15,14 +15,34 @@
  */
 package io.gravitee.rest.api.service.impl;
```
Rename this class to `ApiDuplicatorService_RecalculateApiDefinitionIdsTest` and put it out of the `impl` package
Or
Move all others tests classes in `impl` package 😉

marcambier replied:
As discussed together, the aim is to unit test internal methods without setting them public.
So we made them protected, and the Test class has to be in the same package as the source one in order to call it.

So, I renamed this file and moved all services tests to the impl package, as they all test implementations.

phiz71 commented on gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/rest/api/service/impl/promotion/PromotionServiceTest.java:
```diff
@@ -13,16 +13,21 @@
  * limitations under the License.
  */
-package io.gravitee.rest.api.service.promotion;
+package io.gravitee.rest.api.service.impl.promotion;
```
why ?

marcambier replied:
As discussed together, the aim is to unit test internal methods without setting them public.
So we made them protected, and the Test class has to be in the same package as the source one in order to call it.

So, I moved all services tests to the impl package, as they all test implementations.

phiz71 commented on gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/PageServiceImpl.java:
```diff
@@ -777,10 +782,6 @@ private PageEntity createPage(String apiId, NewPageEntity newPageEntity, String
             }
 
-            if (SYSTEM_FOLDER.equals(newPageType)) {
-                assertNoSystemFolderExists(environmentId, apiId);
```
this prevents someone to create a system folder using the management API, I think this test should remain

marcambier replied:
As discussed together, this exception thrown during import could, in some edge cases, block the user from promoting.
So we changed this behavior to an ERROR message, saying the it will update the existing system folder instead of creating a new one.

phiz71 commented on gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/rest/api/service/PlanService_UpdateTest.java:
```diff
@@ -90,12 +93,17 @@
         when(plan.getStatus()).thenReturn(Plan.Status.STAGING);
         when(plan.getType()).thenReturn(Plan.PlanType.API);
         when(plan.getSecurity()).thenReturn(Plan.PlanSecurityType.API_KEY);
-        when(plan.getValidation()).thenReturn(Plan.PlanValidationType.AUTO);
```
why ?

marcambier replied:
cause this is now an unecessary mockito stubbing.
Before our changes it was called for an inline object mapping. But now this mapping is dedicated to the PlanConverter.

phiz71 commented on gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/promotion/PromotionServiceImpl.java:
```diff
@@ -237,42 +242,35 @@ public PromotionEntity createOrUpdate(PromotionEntity promotionEntity) {
+            if (PromotionStatus.ACCEPTED.equals(promotion.getStatus())) {
                 ApiEntity promoted = null;
 
                 // FIXME: All the methods should take then env id as input instead of relying on GraviteeContext.getCurrentEnv
                 GraviteeContext.setCurrentEnvironment(environment.getId());
```
FIXME ?

marcambier replied:
This FIXME was already existing, and will be fixed with https://github.com/gravitee-io/issues/issues/6071

Commit:
test: cleanup GraviteeContext before and after tests executions

Commit:
style: apply prettier style
